k3s-io/k3s
# Add ability to perform an etcd on-demand snapshot via cli

Signed-off-by: Brian Downs <brian.downs@gmail.com>




#### Proposed Changes ####



This PR adds a new sub-command to the K3s command called `etcd-snapshot` and it takes a number of the standard K3s flags as well as specific flags like `name`, used to give the snapshot a specific name and `dir` used to tell K3s where to save the snapshot. The former use applicable in this use case as well as for cron scheduled snapshots and keeps in like with RKE features for snapshotting and will ultimately be pulled into RKE2. An on-demand snapshot will not effect snaptshot retention.

#### Types of Changes ####



#### Verification ####



1. Start K3s using Etcd and wait for it to get to steady state. 
2. Run `k3s etcd-snapshot` to perform an etcd snapshot.
  A. Verify the snapshot has been performed successfully via log output `{"level":"info","msg":"saved","path":"/var/lib/rancher/k3s/server/db/snapshots/on-demand-<unix-timestamp>"}`
  B. Verify that the newly created snapshot is indeed at the expected location
3. Run `k3s etcd-snapshot --name=my_snapshot` to perform and etcd snapshot now.
  A. Perform the same verification steps from above.

#### Linked Issues ####



#2758 

#### Further Comments ####
